### PR TITLE
feat: add stats logging

### PR DESCRIPTION
**Pull request by jczhong84**
Adding stats logging support. 

Metrics:
  - Number of active users
  - Number of API requests
  - Latency of API requests
  - Number of websocket connections
  - Number of sql session failures
  - Number of scheduled system task failures
  - Number of scheduled datadoc failures
  - Latency of Redis operations
  - Number of query executions

### CODE REVIEW

File: querybook/server/app/auth/__init__.py [hunk 1]
```diff
@@ -1,14 +1,32 @@
...
 from const.path import BUILD_PATH
 from const.datasources import DS_PATH
 from env import QuerybookSettings
 from lib.utils.import_helper import import_module_with_default
+from lib.stats_logger import stats_logger, ACTIVE_USER_COUNTER
 
 
 auth = None
 login_config = None
+active_users = {}
```
Comment (czgu) [inline querybook/server/app/auth/__init__.py:15]: this wouldn't work if you have multiple servers, unless you plan to store this stat in redis

File: querybook/server/app/auth/__init__.py [hunk 2]
```diff
@@ -1,14 +1,32 @@
...
+
+
+def update_active_users():
+    if current_user.id not in active_users:
+        stats_logger.incr(ACTIVE_USER_COUNTER)
+
+    active_users[current_user.id] = time.time()
+
+    for user_id, last_activity_time in list(active_users.items()):
+        if time.time() - last_activity_time > ACTIVE_USER_SESSION_TIMEOUT:
```
Comment (czgu) [inline querybook/server/app/auth/__init__.py:27]: maybe we don't need to log the num of active users? 3 reasons:
1. it is quite annoying to handle with multiple servers
2. the stat is not as useful as number of requests which we also have
3. the num of user can also be extracted from impression log
4. this call is not real time, which pads a variable processing time to the request

File: querybook/server/app/datasource.py [hunk 3]
```diff
@@ -56,6 +61,11 @@ def wrapper(fn):
         @flask_app.route(r"%s%s" % (DS_PATH, url), methods=methods)
         @functools.wraps(fn)
         def handler(**kwargs):
+            # increment the number of api request counter
+            stats_logger.incr(API_REQUEST_COUNTER.format(fn.__name__))
```
Comment (czgu) [inline querybook/server/app/datasource.py:65]: should we use the url here? it is generic right

File: querybook/server/datasources_socketio/helper.py [hunk 4]
```diff
@@ -18,7 +19,13 @@ def handler(*args, **kwargs):
             if not current_user.is_authenticated:
                 LOG.error("Unauthorized websocket access")
                 disconnect()
+                # decrement ws connections counter on disconnect
+                stats_logger.decr(WS_CONNECTIONS_COUNTER)
             else:
+                # decrement ws connections counter on disconnect
+                if url == "disconnect":
+                    stats_logger.decr(WS_CONNECTIONS_COUNTER)
```
Comment (czgu) [inline querybook/server/datasources_socketio/helper.py:27]: do the logging after disconnect function is done?

File: querybook/server/tasks/all_tasks.py [hunk 5]
```diff
@@ -54,3 +59,12 @@ def configure_workers(sender=None, conf=None, **kwargs):
         clean_up_query_execution()
     else:
         LOG.info(f"Starting DEV Celery worker: {sender}")
+
+
+@task_failure.connect
+def handle_task_failure(sender, signal, *args, **kwargs):
+    task_name = sender.name
+    if task_name.startswith("tasks.run_datadoc"):
```
Comment (czgu) [inline querybook/server/tasks/all_tasks.py:67]: there is a task_type field for task and it is either "prod" or "user". Can you make the two connect?

File: querybook/server/app/datasource.py (hunk 3 above)
Comment (jczhong84) [reply querybook/server/app/datasource.py:65]: url is something like `/query_execution/<int:query_execution_id>/`, which will include the parameter types.

**Review by czgu — APPROVED** [review]

**Pull request merged**